DOCSP-30350 Java Write Exception Troubleshooting (#710)
mongodb/docs-java: source/crud/bulk.txt

in an error, the other operations are executed because the write operation is
unordered.

.. _java-bulk-insert-troubleshooting:

Troubleshooting
---------------

Bulk Write Exception
~~~~~~~~~~~~~~~~~~~~

If the driver encounters an error during a bulk write operation, the driver
throws a `MongoBulkWriteException.
<{+core-api+}/MongoBulkWriteException.html>`__ A ``MongoBulkWriteException``
contains a ``writeErrors`` field consisting of a list of one or more
``WriteError`` objects associated with the same bulk write operation. 

> Consider a collection that has a rule where the value of the ``quantity`` field

Reviewer: I think it's useful to emphasise that we're talking about "schema validation rule"

in an error, the other operations are executed because the write operation is
unordered.

.. _java-bulk-insert-troubleshooting:

Troubleshooting
---------------

.. include:: /includes/crud/write-error.rst

Bulk Write Exception
~~~~~~~~~~~~~~~~~~~~

The driver creates a write exception when it creates a ``WriteError`` object. If an
error is detected during a bulk write operation, the driver throws a `MongoBulkWriteException
<{+core-api+}/MongoBulkWriteException.html>`__.

A ``MongoBulkWriteException`` contains a ``writeErrors`` field consisting of a
list of one or more ``WriteError`` objects associated with the same bulk write
operation. 

> Example
> ```````
> 
> For example, the driver throws a ``MongoBulkWriteException`` if your bulk insert
> contains two documents that violate the collection's schema
> validation rules. Suppose the collection has a rule where the value of the
> ``quantity`` field must be an ``int`` type. If your bulk insert contains a
> document with a ``quantity`` of ``"three"`` and another with a ``quantity`` of
> ``"ten"``, the driver prints the following error message:

Reviewer 2: Instead of a sub-sub section, I think we can naturally weave this example in. Included some suggestions for brevity, feel free to tweak.

```suggestion
Consider a collection that has a rule where the value of the
``quantity`` field must be of the ``int`` type. The following example ``MongoBulkWriteException`` is thrown when you attempt to insert a document with a ``quantity`` field value of ``"three"`` and another with a ``quantity`` field value of ``"ten"``.
```

   ... at
   BulkWriteMultipleValidationErrorsExample.main(BulkWriteMultipleValidationErrorsExample.java:30)
   Caused by: com.mongodb.MongoWriteException: WriteError{code=121,
   message='Document failed validation', details={ operator: "$jsonSchema",
   schemaRules: { bsonType: "int", description: "must be an integer" },
   offendingDocument: {"name":"Apple","quantity":"three"} }} at
   com.mongodb.internal.connection.WriteResultHelper.createWriteException(WriteResultHelper.java:50)
   at com.mongodb.internal.connection.ProtocolHelper.getBulkWriteException(ProtocolHelper.java:254)
   ... 19 more 
   Caused by: com.mongodb.MongoWriteException: WriteError{code=121,
   message='Document failed validation', details={ operator: "$jsonSchema",
   schemaRules: { bsonType: "int", description: "must be an integer" },
   offendingDocument: {"name":"Banana","quantity":"ten"} }} at
   com.mongodb.internal.connection.WriteResultHelper.createWriteException(WriteResultHelper.java:50)
   at
   com.mongodb.internal.connection.ProtocolHelper.getBulkWriteException(ProtocolHelper.java:254)
   ... 19 more    

> In the preceding error message, the ``MongoBulkWriteException`` object provides
> a vague description of the error, as documents associated with the same bulk
> operation could produce different error types. Instead, refer to the
> ``WriteError`` objects' ``message`` fields to determine the cause of each error.
> Refer to the ``WriteError`` objects' ``details`` field for more specific
> information, such as the schema rules and the offending documents. To address
> this error, you must either revise the document to adhere to the schema
> validation rules or bypass validation.

Reviewer 2: How do you feel about linking out to the API docs here for the sake of brevity? ([MongoBulkWriteException link](https://mongodb.github.io/mongo-java-driver/5.5/apidocs/driver-core/com/mongodb/MongoBulkWriteException.html), [WriteError link](https://mongodb.github.io/mongo-java-driver/5.5/apidocs/driver-core/com/mongodb/WriteError.html)).

```suggestion
To learn more about the ``MongoBulkWriteException`` and ``WriteError`` types, see MongoBulkWriteException and WriteError in the API documentation.
```
